File: kproc.c
#include "kproc.h"
#include "kmem.h"
#include "uart.h"
#include <stdbool.h>

#define MAXPROCS 1024

#define NCPU 1

enum {
EMPTY = 0,
RUNNING = 1,
IDLE = 2,
};

struct kproc_context sched_ctx;

// this saves all the register state that must be restored when an execption occurs.
// Including sp and lr, because we might end up switching to a different
// (user-space) process after a timer interrupt fires. E.g. after a timer
// interrupt, we won't do a ret to the .
struct exception_frame {
    uint32_t r[15]; // r0-r15; includes sp, lr
};

struct kproc_info {
    int status;
    struct kproc_context ctx;
    uint32_t stack;
};

static struct {
    int curr;
    struct kproc_info procs[MAXPROCS];
    uint32_t nprocs;
} kprocs;

extern char kproc_start[];

void kproc_init()
{
    kprocs.nprocs = 0;
    for (int i = 0; i < MAXPROCS; i++) {
        kprocs.procs[i].status = EMPTY;
    }
}

void kproc_create_thread(void (*fn)(uint32_t), uint32_t args)
{
    if (kprocs.nprocs >= MAXPROCS) {
        uart_puts("Unable to create a thread, reached max procs already\r\n");
        return;
    }

    uint32_t i = kprocs.nprocs;
    kprocs.nprocs += 1;

    // HACK: see kproc_start.S
    // load fn and args into the first two callee-saved registers
    kprocs.procs[i].ctx.r[0] = (uint32_t)fn; // in 64-bit, x19 = fn
    kprocs.procs[i].ctx.r[1] = args; // in 64-bit, x20 = args
    kprocs.procs[i].ctx.lr = (uint32_t)kproc_start;
    kprocs.procs[i].stack = (uint32_t)kmem_alloc();
    kprocs.procs[i].ctx.sp = kprocs.procs[i].stack + PGSIZE; // top of stack, since it grows down
    kprocs.procs[i].ctx.fp = kprocs.procs[i].ctx.sp;
    kprocs.procs[i].status = IDLE;
}

void kproc_scheduler(uint32_t cpu)
{
    bool took_step;
    do {
        took_step = false;
        for (int i = 0; i < MAXPROCS; i++) {
            if (kprocs.procs[kprocs.curr].status == IDLE) {
                kprocs.procs[kprocs.curr].status = RUNNING;

                // set the current stack for the CPU, so that an exception will
                // put stuff on the right stack
                kproc_switch(&sched_ctx, &(kprocs.procs[kprocs.curr].ctx));

                if (kprocs.procs[kprocs.curr].status == EMPTY) {
                    kmem_free((uint8_t*)kprocs.procs[kprocs.curr].stack); // cleanup
                }
                took_step = true;
            }

            kprocs.curr += 1;
            if (kprocs.curr == MAXPROCS) {
                kprocs.curr = 0;
            }
        }
    } while (took_step);
    uart_puts("Took no steps this round; kproc scheduler stopping\r\n");
}

void kproc_yield()
{
    kprocs.procs[kprocs.curr].status = IDLE;
    kproc_switch(&kprocs.procs[kprocs.curr].ctx, &sched_ctx);
}

void kproc_exit()
{
    kprocs.procs[kprocs.curr].status = EMPTY;
    // kmem_free((uint8_t*)kprocs.procs[kprocs.curr].stack); // (fixed BUG): freeing the same stack we're using!
    // kmem_free tries to return, but the stack that it's using has been freed
    kproc_switch(&kprocs.procs[kprocs.curr].ctx, &sched_ctx);
}

File: kproc.zig
const uart = @import("uart.zig");
const panic = @import("panic.zig");
const kmem = @import("kmem.zig");

// Multitasking approach.
// 1) "kernel procs" run in kernel address space and are cooperatively scheduled.
// `kproc.yield()` will yield control.
// 2) user procs that each run in a unique address space and are preempted.

// FIXME: move to arch folder
// callee-saved things that must be saved on kproc_switch()
const CooperativeContext = struct {
    lr: u64,
    sp: u64,
    fp: u64,
    x: [29 - 19 + 1]u64, // x19-x29 callee-saved
    v: [15 - 8 + 1]u64, // v8-x15 callee saved
};

const max = 1024;

const Process = struct {
    ctx: CooperativeContext,
    state: enum { inactive, idle, running, cleanup } = .inactive,

    // The process owns a stack page. Later this might become multiple pages etc.
    // The process must never overflow the stack, or unsafe behavior can occur.
    // This includes the stack space used when calling yield() to save
    // caller-saved registers.
    stack_addr: usize,
};

// TODO: make this a growable array.
var procs: [max]Process = undefined;

var nproc: u64 = 0; // number of currently active processes

const Func = fn (_: u64) void;

pub fn init() void {
    uart.printf("Procs using up {X} bytes of memory\n", .{@sizeOf(@TypeOf(procs))});
}

pub fn spawn(f: *const Func, args: u64) void {
    if (nproc == max) {
        panic.panic("ran out of kprocs\n");
    }
    for (&procs) |*proc| {
        if (proc.state == .inactive) {
            // found a proc, can return
            proc.ctx.x[0] = @intFromPtr(f); // in 64-bit, x19 = fn
            proc.ctx.x[1] = args; // in 64-bit, x20 = args
            proc.state = .idle;
            proc.stack_addr = kmem.alloc_or_panic();
            proc.ctx.sp = proc.stack_addr + kmem.pgsize; // top of stack, since it grows down
            proc.ctx.fp = proc.ctx.sp;
            proc.ctx.lr = @intFromPtr(&kproc_start);
            nproc += 1;
            return;
        }
    }
    panic.panic("unreachable\n");
}

var cur_proc: *Process = undefined;
var sched_ctx: CooperativeContext = undefined;

pub fn schedulerLoop() void {
    var took_step = true;
    while (took_step) {
        took_step = false;
        for (&procs) |*proc| {
            if (proc.state == .idle) {
                proc.state = .running;
                // run the proc
                cur_proc = proc;
                kproc_switch(&sched_ctx, &proc.ctx);

                // XXX: exit() will set the proc to cleanup to indicate that it
                // needs to be cleaned up.
                if (proc.state == .cleanup) {
                    kmem.free(proc.stack_addr); // cleanup
                    proc.state = .inactive;
                } else {
                    proc.state = .idle;
                }
            }
            took_step = true;
        }
    }
    uart.puts("No procs to run\n");
}

pub fn yield() void {
    kproc_switch(&cur_proc.ctx, &sched_ctx);
}

comptime {
    @export(exit, .{ .name = "kproc_exit", .linkage = .strong });
}

fn exit() callconv(.C) void {
    nproc -= 1;
    cur_proc.state = .cleanup;
    uart.puts("process exiting\n");
    yield();
}

extern fn kproc_switch(old: *CooperativeContext, new: *CooperativeContext) void;
extern fn kproc_start() void;

pub extern fn trigger_undef() void;
